packages/marketprice/src/alarms/mod.rs

```diff
@@ -77,136 +75,241 @@ impl<'a> Prefixer<'a> for AlarmStore {
     }
 }
 
-struct AlarmsIndexes<'a>(MultiIndex<'a, AlarmStore, AlarmStore, Addr>);
+struct AlarmsIndexes(MultiIndex<'static, AlarmStore, AlarmStore, Addr>);
 
-impl<'a> IndexList<AlarmStore> for AlarmsIndexes<'a> {
-    fn get_indexes(&'_ self) -> Box<dyn Iterator<Item = &'_ dyn Index<AlarmStore>> + '_> {
-        let v: Vec<&dyn Index<AlarmStore>> = vec![&self.0];
-        Box::new(v.into_iter())
+impl IndexList<AlarmStore> for AlarmsIndexes {
+    fn get_indexes(&self) -> Box<dyn Iterator<Item = &dyn Index<AlarmStore>> + '_> {
+        Box::new(vec![&self.0 as &_].into_iter())
     }
 }
 
-impl<'m> PriceAlarms<'m> {
-    pub const fn new(
-        alarms_below_namespace: &'m str,
-        index_below_namespace: &'m str,
-        alarms_above_namespace: &'m str,
-        index_above_namespace: &'m str,
-    ) -> PriceAlarms<'m> {
-        PriceAlarms {
-            alarms_below_namespace,
-            index_below_namespace,
-            alarms_above_namespace,
-            index_above_namespace,
-        }
-    }
+type IndexedMap = CwIndexedMap<'static, Addr, AlarmStore, AlarmsIndexes>;
 
-    pub fn add_alarm_below<C, BaseC>(
-        &self,
-        storage: &mut dyn Storage,
-        addr: &Addr,
-        alarm: Price<C, BaseC>,
-    ) -> Result<(), AlarmError>
-    where
-        C: Currency,
-        BaseC: Currency,
-    {
-        Ok(self
-            .alarms_below()
-            .save(storage, addr.to_owned(), &AlarmStore::new(&alarm))?)
-    }
+fn alarms_index(alarms_namespace: &'static str, index_namespace: &'static str) -> IndexedMap {
+    let indexes = AlarmsIndexes(MultiIndex::new(
+        |_, price| price.to_owned(),
+        alarms_namespace,
+        index_namespace,
+    ));
 
-    pub fn add_alarm_above_or_equal<C, BaseC>(
-        &self,
-        storage: &mut dyn Storage,
-        addr: &Addr,
-        alarm: Price<C, BaseC>,
-    ) -> Result<(), AlarmError>
-    where
-        C: Currency,
-        BaseC: Currency,
-    {
-        Ok(self
-            .alarms_above_or_equal()
-            .save(storage, addr.to_owned(), &AlarmStore::new(&alarm))?)
-    }
+    IndexedMap::new(alarms_namespace, indexes)
+}
+
+pub struct PriceAlarms<'storage, S>
+where
+    S: Deref<Target = dyn Storage + 'storage>,
```

Comment:
[everywhere] why have you decided to switch from `&'a dyn Storage` to `Deref<Target = dyn Storage + 'a>`

Reply:
we do not need to work with other types than `&'a dyn Storage`

Reply:
We do. I did it so I don't have to separate the the implementations for immutable and mutable versions, depending on the storage.

Reply:
Reasoning is that, `&_` implements `Deref` but `&mut _` implements both `Deref` and `DerefMut`.

Reply:
That difference can be seen on line +182

Reply:
indeed we need to use &mut as well